File: mobile/android/veripaykit/src/main/java/com/veripay/sdk/ChallengeNonce.kt
package com.veripay.sdk

import java.util.UUID

data class ChallengeNonce(
    val value: String,
    val expiresAtMillis: Long
)

class SingleUseNonceStore(
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
    private val nonceFactory: () -> String = { UUID.randomUUID().toString() }
) {
    private val expirations = mutableMapOf<String, Long>()

    fun issue(ttlMillis: Long = 90_000): ChallengeNonce {
        require(ttlMillis > 0) { "TTL must be positive" }
        val value = nonceFactory()
        val expiresAt = nowMillis() + ttlMillis
        expirations[value] = expiresAt
        return ChallengeNonce(value, expiresAt)
    }

    fun consume(nonce: ChallengeNonce): Boolean {
        val expiration = expirations.remove(nonce.value) ?: return false
        return expiration > nowMillis()
    }
}

File: mobile/ios/Sources/VeriPayKit/ChallengeNonce.swift
import Foundation

public struct ChallengeNonce: Equatable {
    public let value: String
    public let expiresAt: Date

    public init(value: String, expiresAt: Date) {
        self.value = value
        self.expiresAt = expiresAt
    }
}

public final class SingleUseNonceStore {
    private var expirations: [String: Date] = [:]
    private let now: () -> Date
    private let nonceFactory: () -> String

    public init(
        now: @escaping () -> Date = Date.init,
        nonceFactory: @escaping () -> String = { UUID().uuidString }
    ) {
        self.now = now
        self.nonceFactory = nonceFactory
    }

    public func issue(ttl: TimeInterval = 90) -> ChallengeNonce {
        let nonce = nonceFactory()
        let expiresAt = now().addingTimeInterval(ttl)
        expirations[nonce] = expiresAt
        return ChallengeNonce(value: nonce, expiresAt: expiresAt)
    }

    public func consume(_ challenge: ChallengeNonce) -> Bool {
        guard let expiration = expirations.removeValue(forKey: challenge.value) else {
            return false
        }
        return expiration > now()
    }
}
